File: engine/language_client_ruby/test/type_builder_test.rb
require 'test/unit'
require 'baml'

# this test suite verifies that our type builder system correctly handles
# string representations of complex types like classes and enums. this is
# important for debugging and logging purposes, as it helps
# understand the structure of their type definitions at runtime.
class TypeBuilderTest < Test::Unit::TestCase

  # tests that class definitions are properly stringified with all their
  # properties and metadata intact. this helps ensure our type system
  # maintains semantic meaning when displayed to users.
  def test_class_string_representation
    # start with a fresh type builder - this is our main interface
    # for constructing type definitions programmatically
    builder = Baml::Ffi::TypeBuilder.new

    # create a new user class - this represents a person in our system
    # with various attributes that describe them
    user_class = builder.class_('User')

    # define the core properties that make up a user profile
    # we use aliases and descriptions to make the api more human-friendly
    user_class.property('name')
      .alias('username')  # allows 'username' as an alternative way to reference this
      .description('The user\'s full name')  # helps explain the purpose

    user_class.property('age')
      .description('User\'s age in years')  # clarifies the expected format

    user_class.property('email')  # sometimes a property name is self-explanatory

    # convert our type definition to a human-readable string
    # this is invaluable for debugging and documentation
    output = builder.to_s
    puts "\nClass output:\n#{output}\n"

    # verify that the string output matches our expectations
    # we check for key structural elements and metadata
    assert_match(/TypeBuilder\(\n  Classes: \[\n    User \{/, output)
    assert_match(/name \(unknown-type\) \(alias=String\("username"\), description=String\("The user's full name"\)\)/, output)
    assert_match(/age \(unknown-type\) \(description=String\("User's age in years"\)\)/, output)
    assert_match(/email \(unknown-type\)/, output)
  end

  # tests that enum definitions are correctly stringified with their
  # values and associated metadata. enums help us model fixed sets
  # of options in a type-safe way.
  def test_enum_string_representation
    # create a fresh builder for our enum definition
    builder = Baml::Ffi::TypeBuilder.new

    # define a status enum to track user account states
    # this gives us a type-safe way to handle different user situations
    status_enum = builder.enum('Status')

    # add the possible status values with helpful metadata
    # active users are currently using the system
    status_enum.value('ACTIVE')
      .alias('active')  # lowercase alias for more natural usage
      .description('User is active')  # explains the meaning

    # inactive users have temporarily stopped using the system
    status_enum.value('INACTIVE')
      .alias('inactive')

    # pending users are in a transitional state
    status_enum.value('PENDING')

    # generate a readable version of our enum definition
    output = builder.to_s
    puts "\nEnum output:\n#{output}\n"

    # verify the string representation includes all our carefully
    # defined values and their metadata
    assert_match(/TypeBuilder\(\n  Enums: \[\n    Status \{/, output)
    assert_match(/ACTIVE \(alias=String\("active"\), description=String\("User is active"\)\)/, output)
    assert_match(/INACTIVE \(alias=String\("inactive"\)\)/, output)
    assert_match(/PENDING/, output)
  end
end

File: engine/language_client_typescript/__test__/type_builder.test.ts
// import the typescript wrapper for the type builder that provides a clean interface
// over the native rust implementation
import { TypeBuilder } from '../native';

describe('TypeBuilder', () => {
  // test that we can create classes with properties and add metadata like aliases and descriptions
  it('should provide string representation for classes with properties and metadata', () => {
    // create a fresh type builder instance to work with
    const builder = new TypeBuilder();

    // get a reference to a class named 'user', creating it if needed
    const userClass = builder.getClass('User');

    // add properties to the user class with helpful metadata
    // the name property has both an alias and description
    userClass.property('name')
      .alias('username')                     // allows referencing the property as 'username'
      .description('the user\'s full name'); // explains what this property represents

    // age property just has a description
    userClass.property('age')
      .description('user\'s age in years');  // clarifies the age units

    // email is a basic property with no extra metadata
    userClass.property('email');             // simple email field

    // convert all the type definitions to a readable string
    const output = builder.toString();

    // make sure the output has the expected class structure
    expect(output).toContain('TypeBuilder(\n  Classes: [\n    User {');
    // verify each property appears with its metadata
    expect(output).toContain("name (unknown-type) (alias=String(\"username\"), description=String(\"the user's full name\"))");
    expect(output).toContain("age (unknown-type) (description=String(\"user's age in years\"))");
    expect(output).toContain('email (unknown-type)');
  });

  // test that we can create enums with values and add metadata like aliases and descriptions
  it('should provide string representation for enums with values and metadata', () => {
    // create a fresh type builder instance to work with
    const builder = new TypeBuilder();

    // get a reference to an enum named 'status', creating it if needed
    const statusEnum = builder.getEnum('Status');

    // add possible values to the status enum with helpful metadata
    // active state has both an alias and description
    statusEnum.value('ACTIVE')
      .alias('active')                    // allows using lowercase 'active'
      .description('user is active');     // explains what active means

    // inactive state just has an alias
    statusEnum.value('INACTIVE')
      .alias('inactive');                 // allows using lowercase 'inactive'

    // pending is a basic value with no extra metadata
    statusEnum.value('PENDING');          // simple pending state

    // convert all the type definitions to a readable string
    const output = builder.toString();

    // make sure the output has the expected enum structure
    console.log(output);
    expect(output).toContain(`TypeBuilder(
  Enums: [
    Status {`);
    // verify each value appears with its metadata
    expect(output).toContain('ACTIVE (alias=String(\"active\"), description=String(\"user is active\"))');
    expect(output).toContain('INACTIVE (alias=String(\"inactive\"))');
    expect(output).toContain('PENDING');
  });
});